src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType.DelegateTemplateSymbol.cs

```diff
@@ -0,0 +1,178 @@
+                // https://github.com/dotnet/roslyn/issues/56808: Synthesized delegates should include BeginInvoke() and EndInvoke().
+                var invokeMethod = createInvokeMethod(this, refKinds, voidReturnTypeOpt);
+                _members = ImmutableArray.Create<Symbol>(constructor, invokeMethod);
+
+                static SynthesizedDelegateInvokeMethod createInvokeMethod(AnonymousDelegateTemplateSymbol containingType, RefKindVector refKinds, TypeSymbol? voidReturnTypeOpt)
+                {
+                    var typeParams = containingType.TypeParameters;
+
+                    int parameterCount = typeParams.Length - (voidReturnTypeOpt is null ? 1 : 0);
+                    var parameterTypes = ArrayBuilder<TypeWithAnnotations>.GetInstance(parameterCount);
+                    var parameterRefKinds = ArrayBuilder<RefKind>.GetInstance(parameterCount);
+                    for (int i = 0; i < parameterCount; i++)
+                    {
+                        parameterTypes.Add(TypeWithAnnotations.Create(typeParams[i]));
+                        parameterRefKinds.Add(refKinds.IsNull ? RefKind.None : refKinds[i]);
+                    }
+
+                    // if we are given Void type the method returns Void, otherwise its return type is the last type parameter of the delegate:
+                    var returnType = TypeWithAnnotations.Create(voidReturnTypeOpt ?? typeParams[parameterCount]);
+                    var returnRefKind = (refKinds.IsNull || voidReturnTypeOpt is { }) ? RefKind.None : refKinds[parameterCount];
+
+                    var method = new SynthesizedDelegateInvokeMethod(containingType, parameterTypes, parameterRefKinds, returnType, returnRefKind);
+                    parameterRefKinds.Free();
+                    parameterTypes.Free();
+                    return method;
+                }
+
+                static ImmutableArray<TypeParameterSymbol> createTypeParameters(AnonymousDelegateTemplateSymbol containingType, int parameterCount, bool returnsVoid)
+                {
+                    var typeParameters = ArrayBuilder<TypeParameterSymbol>.GetInstance(parameterCount + (returnsVoid ? 0 : 1));
+                    for (int i = 0; i < parameterCount; i++)
+                    {
+                        typeParameters.Add(new AnonymousTypeManager.AnonymousTypeParameterSymbol(containingType, i, "T" + (i + 1)));
+                    }
+
+                    if (!returnsVoid)
+                    {
+                        typeParameters.Add(new AnonymousTypeManager.AnonymousTypeParameterSymbol(containingType, parameterCount, "TResult"));
+                    }
+
+                    return typeParameters.ToImmutableAndFree();
+                }
+            }
+
+            internal AnonymousDelegateTemplateSymbol(AnonymousTypeManager manager, AnonymousTypeDescriptor typeDescr, ImmutableArray<TypeParameterSymbol> typeParametersToSubstitute)
```

Comment:
nit: xml doc on these constructors would be useful. This constructor is for non-generic case, and above is generic case